pkg/reconciler/revision/revision.go

```diff
@@ -129,13 +133,14 @@ func (c *Reconciler) reconcileDigest(ctx context.Context, rev *v1.Revision) erro
 	}
 	digestGrp.Wait()
 	close(digests)
+	// If there is a digest error no need to create a containerStatus object
+	if d.digestError != nil {
```

Comment:
This is a little weird indeed. One would expect this to be the usecase of the errgroup, but I get you need more state than just the error.

What about this:
1. Remove the redundant error check here.
2. Build ContainerStatuses below outside of the struct.
3. As soon as the loop below hits an error, exit early.
4. If the foor loop below passes just fine, assign the ContainerStatuses to the struct.

Pseudo-code:

```golang
var (
  servingDigest string
  containerStatuses = make([]v1.ContainerStatuses, 0, len(digests))
)
for v := range digests {
  if v.digestError != nil {
    rev.Status.MarkContainerHealthyFalse(v1.ReasonContainerMissing,
      v1.RevisionContainerMissingMessage(
        v.image, v.digestError.Error()))
    return v.digestError
  }
  if v.isServingContainer {
    servingDigests = v.digestValue
  }
  containerStatuses = append(containerStatuses, v1.ContainerStatuses{
    Name:        v.containerName,
    ImageDigest: v.digestValue,
  })
}
// If we reached here, no errors happened.
rev.Status.DeprecatedImageDigest = servingDigest
rev.Status.ContainerStatuses = containerStatuses
```

That should allow the same semantics without the extra variable and mutex above.

Reply:
yeah also `errgroup.Wait()` returns an error

Reply:
@markusthoemmes Thank you updated PR

@dprotaso the [errgroup](https://github.com/knative/serving/pull/7801/files#diff-bdad1b2ce7ce2630853dc1279891c246R130) returns nil so that's why not verifying error as part of [errgroup.Wait()](https://github.com/knative/serving/pull/7801/files#diff-bdad1b2ce7ce2630853dc1279891c246R133)